cidc_api/resources/upload_jobs.py

```diff
@@ -86,18 +87,67 @@ def get_upload_job(upload_job: UploadJobs):
     return upload_job
 
 
+def requires_upload_token_auth(get_upload_id):
+    """
+    Decorator factory that adds "upload token" authentication to an endpoint. 
+    `get_upload_id` takes the arguments passed to the decorated endpoint
+    and extracts the ID of the upload job to authenticate against.
+    If the requesting user provided the correct upload token for the upload job, 
+    then consider them authenticated.
+
+    Requests must provide the upload token as a URL query param `token`.
+    """
+
+    check_id_token = requires_auth("[upload token auth - no RBAC possible]")(
+        lambda: None
+    )
+    token_schema = Schema.from_dict({"token": fields.Str(required=True)})(
+        unknown=INCLUDE
+    )
+
+    def decorator(endpoint):
+        # Flag this endpoint as authenticated
+        endpoint.is_protected = True
+
+        @wraps(endpoint)
+        @use_args(token_schema, location="query")
+        def wrapped(args, *pos_args, **kwargs):
+            # Try to get the user associated with this request
+            try:
+                check_id_token()
+                user = get_current_user()
+            except:
+                user = None
+
+            token = args["token"]
+            upload_id = get_upload_id(*pos_args, **kwargs)
+            upload_job = UploadJobs.find_by_id(upload_id)
+
+            if not upload_job:
+                if user:
+                    raise NotFound(f"No upload_job with id {upload_id}")
+                else:
+                    raise Unauthorized("upload_job token authentication failed")
+
+            if str(token) != str(upload_job.token):
+                raise Unauthorized("upload_job token authentication failed")
+
+            return endpoint(*pos_args, **kwargs)
+
+        return wrapped
+
+    return decorator
+
+
 @upload_jobs_bp.route("/<int:upload_job>", methods=["PATCH"])
-@requires_auth("upload_jobs", upload_job_roles)
+@requires_upload_token_auth(get_upload_id=lambda *a, **kw: kw["upload_job"])
```

Comment:
nit: but then if this decorator is for upload job only and is very specific to this, why do this generic-ish `get_upload_id`.

So for me it's either should be more abstract or more upload_job specific (like specificities baked-in).